### DISCUSSION TITLE
Is there a way to automatically search through DANDI files?

### DISCUSSION THREAD

**Discussion by mudd54** [dandi/helpdesk#157] [Q&A] [answered]
Hello,
I have derived data that I’m looking to link to the data of these 3 dandisets : https://dandiarchive.org/dandiset/000713?search=neuropixels&pos=2,https://dandiarchive.org/dandiset/000021, https://dandiarchive.org/dandiset/000409?search=IBL&pos=2 .

To do so I need to match my local files with the corresponding nwb hosted on DANDI.  We are planning to use LINDI to do this @magland

It doesn’t seem feasible to search manually for each matching file. **Is there a way to search automatically or programmatically through folders and file names on DANDI?  These files have session ids from the original data but they do not seem to be available on DANDI that is obvious to me.**
Thank you.

**Comment by magland** [comment, marked as answer]
@mudd54 I'm glad you're going to make use of LINDI for this, and I'm happy to assist.

You can use the dandiarchive python API for this
https://dandi.readthedocs.io/en/latest/modref/dandiarchive.html

Let me know if you want me to provide a template script to help you get started.

**Reply by mudd54** [reply]
@magland The API has worked well to find the files I need, thank you!

In regards to LINDI, I'm specifically looking to link TimeIntervals to the ephys time series data of the dandisets I mentioned, but I'm having some trouble doing so. 

**Is it necessary to ensure the derived data has the same reference time object as the original NWB?  The Time Intervals values correspond to the values in the reference timeseries but are not linked to it in any formal way in the object itself.**

Is there a way to do this, maybe using create_dataset method in a staging area? I've been trying to work off the examples listed on the lindi readme.
Thanks.

**Reply by kabilar** [reply]
Hi @rly @bendichter, do you happen to know the answer to the above question?  Thank you.

**Reply by rly** [reply]
> Is it necessary to ensure the derived data has the same reference time object as the original NWB? The Time Intervals values correspond to the values in the reference timeseries but are not linked to it in any formal way in the object itself.

All timestamps within an NWB file must be based on the same clock where the session start time of the NWB file is t=0. 

I think you can use PyNWB to load the original NWB data as LINDI, add rows to a `TimeIntervals` table where the "timeseries" column value is (idx_start, count, timeseries as an object reference), and call `client.write_lindi_file('new.nwb.lindi.json')` to write a new LINDI file.

The workflow for using PyNWB to append derived data to an NWB file represented as a LINDI JSON is not so well established yet, so let us know if you run into issues. We could also set up a Zoom appointment to work through this together. Feel free to [book a time](https://calendly.com/rly1/50-minute-meeting?preview_source=et_card) with me.

(Note: I think sharing LINDI JSON files that represent NWB data on DANDI is not officially supported by DANDI, but DANDI team, please correct me if I am wrong.)

fyi @oruebel

**Reply by AngCamp** [reply]
@rly myself and Mudd have booked time with you tomorrow 3pm-350pm.  See you then and thank you.

**Reply by kabilar** [reply]
Thank you, Ryan.  On the DANDI staging server (and possibly the production server), Neurosift has now been added as an `External Service` to open `.nwb.lindi.json` files. Jeremy has an [example dataset](https://gui-staging.dandiarchive.org/dandiset/213569/draft/files?location=000946%2Fsub-BH494&page=1) on the staging server.  As of yet, I don't think the team has done too much testing with the LINDI JSON files.

**Reply by rly** [reply]
Following up with @AngCamp and @mudd54:

Here are the notebooks I referenced using the DANDI API:
[Simple search](https://github.com/dandi/example-notebooks/blob/master/tutorials/neurodatarehack_2024/simple_dandiset_search.ipynb)
[Advanced search where you open individual files](https://github.com/dandi/example-notebooks/blob/master/tutorials/neurodatarehack_2024/advanced_asset_search.ipynb)

Here is the code demonstrating how to modify an NWB file represented as a LINDI JSON, using PyNWB:
https://github.com/NeurodataWithoutBorders/lindi/pull/90

You could further modify that code to link to a specific probe in your `TimeIntervals` table (a column can contain objects like `ElectrodeGroup` objects). You could also link to a specific row of the electrodes table by doing something like:
```python
# add a column to the users table that references rows of the addresses table
new_table.add_column(
    name='electrode',
    description='the electrode',
    table=nwbfile.electrodes
)

# add rows to the users table
new_table.add_row(
    start_time=1.0,
    stop_time=2.0,
    electrode=5,  # <-- this is a row index, NOT the id
)
```

**Reply by mudd54** [reply]
@rly Thanks so much! Writing to the nwbs seems to be working well.  
One issue I am running into is some of the lindi links for the ABI Visual Behaviour nwbs give me a 404 error, for example: https://lindi.neurosift.org/dandi/dandisets/000713/assets/0b49c6e1-d276-4d92-9f0d-4095b0195a40/nwb.lindi.json while others work well: https://lindi.neurosift.org/dandi/dandisets/000713/assets/7d6328a5-42d3-4627-83d5-25e96769613f/nwb.lindi.json . 

All I've been doing is swapping out the asset id for the nwb I'm interested in. Do you or @magland have some idea why this might be happening? Is there any alternate link format I could use, or are some of the files not publicly accessible through LINDI? Thank you.

**Reply by magland** [reply]
@mudd54 There is an indexer that runs every morning that prepares lindi.json files for new public files on DANDI. It only runs partially for a fixed amount of time, so some may not be prepared yet. But I thought that 000713 was fully covered. Are you sure that asset ID corresponds to an asset that's actually in the current version of the 000713 Dandiset?

**Reply by mudd54** [reply]
@magland  Below is the code I use to get the asset ID. I use dandiarchive to search through assets until an asset path matches my regex, and then I swap the corresponding asset id into the lindi urls. 

I believe the dandiset url that gets parsed is current. Additionally, I've been manually double checking some of the asset IDs by accessing the metadata through dandi on my browser, and the asset IDs in the metadata and what my code returns match up. 

If it is an issue with the indexer, I can wait until the dandiset has been completed. 

```
def get_url_from_abi_visbehave_session_id(session_id_string, ecephys_sessions_df):
    
    '''
    Get the lindi url from the unique session identifier (allen visbehave)

    session_id_string: (str) String of the session id 
    ecephys_sessions_df: Csv of the ecephys metadata for this dandiset

    returns: lindi url corresponding to the dandi file 
    '''

    # dandi file folders are named after mice ids
    session_id_mask = ecephys_sessions_df['ecephys_session_id'] == int(session_id_string)

    dandi_folder = str(ecephys_sessions_df['mouse_id'][session_id_mask].item())
    dandi_doa = ecephys_sessions_df['date_of_acquisition'][session_id_mask].item()
    dandi_file_name_regex = datetime_to_regex(dandi_doa, dandi_folder)

    # get the url of the dandiset folder 
    dandi_folder_url = f'https://dandiarchive.org/dandiset/000713/0.240702.1725/files?location=sub-{dandi_folder}'

    # using the dandiarchive module, navigate through the dandi folder until a file name matches the regex.
    # This is unfortunately necessary because the session start times in the metadata csvs do not 
    # exactly match the session start times that are used to generate file names -_- 
    # looks like a timezone issue because the dates, minutes and seconds align, while the hours do not.
    # alternately, we could open files and check the session id ('identifier') inside, however it will be slow to load 
    parsed_url = dandiarchive.parse_dandi_url(dandi_folder_url)
    dandi_asset_id = str
    dandi_asset_path = str

    with parsed_url.navigate() as (client, dandiset, assets):

        asset_ids = parsed_url.get_asset_ids(client)

        for asset, asset_id in zip(assets,asset_ids):
            if re.match(dandi_file_name_regex, asset.path):
                print('Match found!')
                print(f'Asset path: {asset.path}, Asset ID: {asset_id} for session {session_id_string}')
                dandi_asset_id = asset_id
                dandi_asset_path = asset.path
                break
            
    
    if dandi_asset_id == '':
        print('No match found for the dandi file name')
        return None, None

    # URL of the remote .nwb.lindi.json file
    url = f'https://lindi.neurosift.org/dandi/dandisets/000713/assets/{dandi_asset_id}/nwb.lindi.json'
    dandi_asset_name = dandi_asset_path.split('/')[-1]

    return url, dandi_asset_name
```

**Reply by magland** [reply]
Makes sense. To figure this out, it would be helpful if you could provide the path of the asset with the ID that is not indexed.

**Reply by mudd54** [reply]
@magland Here's a csv file of the ones I'm interested in that have dead links (paths are in the second column).

I've been using the https://lindi.neurosift.org/dandi/dandisets/000713/assets/[asset_id]/nwb.lindi.json url format BTW. Thank you for your assistance with this. 

[000713_dead_links.csv](https://github.com/user-attachments/files/16639670/000713_dead_links.csv)

**Reply by mudd54** [reply]
@magland We are now at the stage where we would like to try uploading our nwbs with additional data as described here https://github.com/magland/lindi-dandi .

What would be the best way to go about uploading staged data to DANDI? It seems that currently lindi does not have the on_upload_blob and on_upload_main functions needed to run LindiH5pyFile.upload() . Can we adapt code from this https://github.com/magland/lindi-dandi/blob/main/devel/lindi_test_2.py file?

Additionally, is there special authentication needed? Would it be possible to obtain an api key for doing this? 
Thank you for all your help.
